solidity/test/token-distribution/TokenDistributorTest.js

```diff
@@ -0,0 +1,283 @@
+      await tokenDistributor.allocate(
+        testData.merkle.merkleRoot,
+        testData.merkle.tokenTotal,
+        { from: owner }
+      )
+
+      await expectRevert(
+        tokenDistributor.allocate("0x1234567890", testData.merkle.tokenTotal, {
+          from: owner,
+        }),
+        "tokens were already allocated"
+      )
+    })
+
+    it("reverts on empty merkle root", async function () {
+      await expectRevert(
+        tokenDistributor.allocate([], testData.merkle.tokenTotal, {
+          from: owner,
+        }),
+        "merkle root cannot be empty"
+      )
+    })
+
+    it("reverts on token transfer failure", async function () {
+      await expectRevert(
+        tokenDistributor.allocate(
+          testData.merkle.merkleRoot,
+          testData.merkle.tokenTotal + 1,
+          {
+            from: owner,
+          }
+        ),
+        "SafeERC20: low-level call failed"
+      )
+    })
+  })
+
+  describe("claim", async () => {
+    beforeEach(async () => {
+      await tokenDistributor.allocate(
+        testData.merkle.merkleRoot,
+        testData.merkle.tokenTotal,
+        { from: owner }
+      )
+    })
+
+    it("destination address signature verification", async function () {
+      const testCases = new Map([
+        [
+          "completes when signed by recipient, submitted by third-party",
+          {
+            signer: recipientIndex,
+            submitter: thirdPartyIndex,
+            expectRevert: false,
+          },
+        ],
+        [
+          "completes when signed by recipient, submitted by recipient",
+          {
+            signer: recipientIndex,
+            submitter: recipientIndex,
+            expectRevert: false,
+          },
+        ],
+        [
+          "reverts when signed by third-party, submitted by recipient",
+          {
+            signer: thirdPartyIndex,
+            submitter: recipientIndex,
+            expectRevert: true,
+          },
+        ],
+        [
+          "completes when signed by recipient, submitted by destination",
+          {
+            signer: recipientIndex,
+            submitter: destinationIndex,
+            expectRevert: false,
+          },
+        ],
+        [
+          "reverts when signed by destination, submitted by recipient",
+          {
+            signer: destinationIndex,
+            submitter: recipientIndex,
+            expectRevert: true,
+          },
+        ],
+        [
+          "reverts when signed by destination, submitted by destination",
+          {
+            signer: destinationIndex,
+            submitter: destinationIndex,
+            expectRevert: true,
+          },
+        ],
+      ])
+
+      for (const [testCaseName, testCaseData] of testCases) {
+        await createSnapshot()
+
+        console.log(`      ${testCaseName}`)
+        try {
+          const signature = web3.eth.accounts.sign(
+            web3.utils.sha3(destination),
```

Comment:
do you think we can use an alias of sha3 - `web3.utils.keccak256` to align these tests with the signature verification on smart contract?

Reply:
Good idea!